### ISSUE TITLE
Array hashing is in Ruby 1.8 mode

### ISSUE THREAD

**Issue by kml** [jruby/jruby#2437]
Different arrays, are hashing very similarly

```
>> [].hash
=> 0
>> [2].hash
=> 0
```

MRI's behaviour is different:

```
>> [].hash
=> 3252255281033687896
>> [2].hash
=> 3208161167426848866
```

**Comment by headius** [comment]
We calculate based on array length and hash of the elements, with some bitshifting etc.

Based on [MRI's implementation](https://gist.github.com/headius/2b65c9bbdc142d811cfe) it appears we're missing a few things:

* Murmur calculation of the hash
* Pseudo-random salt (MRI uses rb_ary_hash pointer address)

**Comment by Who828** [comment]
I can take this up tonight, looks like a nice bug get started.

**Comment by Who828** [comment]
Here is the change, https://gist.github.com/Who828/056848941c6efee4044e
Let me know if that's what you expected before I send the PR.

**Comment by headius** [comment]
The PR looks good, but I wonder if we should also add some kind of pseudo-random salt like MRI. Could be a one-shot random number from almost anywhere. Or we could skip it.

@enebo, @tduehr: Yay or nay? Do we need it?

**Comment by tduehr** [comment]
I think it's a good idea to have one per runtime. I'm not entirely sure the full impact, there may be some issues relating to shared objects between runtimes.

**Comment by headius** [comment]
#2443 gets us using proper hashing algorithm but does not include salting. I'll file a separate issue for that, since I think we need a bit more discussion about what to use for salt.

**Comment by headius** [comment]
Had to revert the patch because it has other issues.

**Comment by enebo** [comment]
The index and realLength passed into the hash was not for the bytelist but for the array.  That was one thing I noticed.  The other issue I saw too was [nil, false].hash nil.toString().getBytes() seems to return null?

**Comment by MSNexploder** [comment]
Looks like this is fixed in 9.1.10.0:
```ruby
[].hash
=> 5632391355580947474
[2].hash
=> 4399786657182928470
```

**Comment by headius** [comment]
Woot, thanks!
